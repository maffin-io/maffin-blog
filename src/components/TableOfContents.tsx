import React from 'react';
import Link from 'next/link';
import { Post } from '@/app/api/getPosts';

export type TableOfContentsProps = {
  post: Post;
};

export default function TableOfContents({ post }: TableOfContentsProps) {
  const { content } = post;
  const headers = collectHeadings(content);

  return (
    <ul className="list-none pl-0">
      {headers.map(h => {
        const href = `#${h.id}`;

        return (
          <li
            key={h.title}
            className={`pl-${3 * (h.level - 1)}`}
          >
            <Link
              href={href}
            >
              {h.title}
            </Link>
          </li>
        );
      })}
    </ul>
  );
}

type Heading = {
  title: string;
  level: number;
  id: string | null;
};

function collectHeadings(content: string): Heading[] {
  const headingsRegex = /<h([1-6])(?:\s+[^>]*)*>(.*?)<\/h[1-6]>/g;
  const headings = [];
  let match;

  while ((match = headingsRegex.exec(content)) !== null) {
    const headingObject = {
      title: match[2],
      level: parseInt(match[1], 10),
      id: getAttributeValue('id', match[0]),
    };
    headings.push(headingObject);
  }

  return headings;
}

// Function to extract attribute value from a string
function getAttributeValue(attribute: string, text: string): string {
  const regex = new RegExp(`${attribute}=['"]([^'"]+)['"]`);
  const match = regex.exec(text);
  return match ? match[1] : '';
}
