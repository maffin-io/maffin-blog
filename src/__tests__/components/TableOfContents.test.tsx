import { render } from '@testing-library/react';

import TableOfContents from '@/components/TableOfContents';
import type { Post } from '@/app/api/getPosts';

describe('TableOfContents', () => {
  let post: Post;
  beforeEach(() => {
    post = {
      content: '<h1 id="id-1">H1</h1><h2 id="id-2">H2</h2>',
    } as Post;
  });

  it('collects headings and renders as expected', () => {
    const { container } = render(<TableOfContents post={post} />);

    expect(container).toMatchSnapshot();
  });

  it('works ok when no id', () => {
    post.content = '<h1>h</h1>';
    const { container } = render(<TableOfContents post={post} />);

    expect(container).toMatchSnapshot();
  });
});
