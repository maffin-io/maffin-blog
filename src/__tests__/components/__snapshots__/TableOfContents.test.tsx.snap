// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableOfContents collects headings and renders as expected 1`] = `
<div>
  <ul
    class="list-none pl-0"
  >
    <li
      class="pl-0"
    >
      <a
        href="#id-1"
      >
        H1
      </a>
    </li>
    <li
      class="pl-3"
    >
      <a
        href="#id-2"
      >
        H2
      </a>
    </li>
  </ul>
</div>
`;

exports[`TableOfContents works ok when no id 1`] = `
<div>
  <ul
    class="list-none pl-0"
  >
    <li
      class="pl-0"
    >
      <a
        href="#"
      >
        h
      </a>
    </li>
  </ul>
</div>
`;
